Add order filters

jakirkham opened the pull request:
Partially addresses issue ( https://github.com/jakirkham/dask-ndfilters/issues/13 ).

Creates a module and tests for order-based filters (e.g. minimum, maximum, median, etc.) for use with Dask.

jakirkham committed:
Add order filter module stub

Create a stub module for order-based filters (e.g. minimum, maximum,
median, etc.). Also add a stub module for testing it, which just imports
from the module for now.

jakirkham committed:
Add a private function to normalize the footprint

As many of the filters take a size or a footprint, try to standardize
the procedure for normalizing this to give us a simple Boolean NumPy
array. Also handle any of the cases where there are issues specifying
these and make it clear to the end user.

jakirkham committed:
Add a private function to normalize the origin

As many of the filters take an origin, try to standardize the origin and
ensure it is well-formed. This value is also important for determining
the halos needed. So it is important that we tidy it up such that the
computation of the halos widths proceeds simply.

jakirkham committed:
Wrapped median filter

This is an initial attempt at providing a wrapped median filter. It
needs a bit more testing. However it should provide a good way of
performing a median filter with appropriate haloing with Dask.

jakirkham committed:
Expose median_filter from the module

jakirkham committed:
Add a workaround for size 0 depths in Dask

Appears that Dask has a bug when size 0 depths are present.
Unfortunately we can't simply skip adding halos in this case as we may
need halos on some dimensions and not others. While we could special
case depths that are 0 on every dimension, this just complicates the
code on our end. So we add an easy workaround that adds a 1 element halo
to all borders no matter what. As this seems to workaround the problem
sufficiently, this should be fine until Dask solves this properly.

jakirkham committed:
Add some tests of the median filter

jakirkham committed:
Refactor a wrapper for ordering filters

This will allow us to extend it to other things beyond the median like
max and min filters.

jakirkham committed:
Customize docstring of wrapping function

jakirkham committed:
Wrap maximum and minimum filters

jakirkham committed:
Expose minimum_filter and maximum_filter at top

jakirkham committed:
Add tests for maximum and minimum filters too

jakirkham committed:
Float origin is not permissible for max and min

The maximum and minimum filter in SciPy won't accept a float origin
unlike the median filter. So don't bother testing those with a float
origin unlike the median filter.

jakirkham commented on dask_ndfilters/order.py:
```diff
@@ -0,0 +1,148 @@
+        origin = ndim * (origin,)
+
+    origin = numpy.array(origin)
+    origin = numpy.fix(origin).astype(int)
```
Not sure whether this is an ok thing to do or whether we should just be erroring out on non-integer values. This is what SciPy ndimage's `median_filter` does, but the `minimum_filter` and `maximum_filter` actually error out for not integer values. Asked SciPy for some clarification on this.

xref: https://github.com/scipy/scipy/issues/7343

jakirkham replied:
Probably the easiest thing is to turn this into an error outright. Can always soften it later if we decide to.

jakirkham replied:
It is now a `TypeError` to use anything other than integral value(s) for the `origin`.

jakirkham committed:
Wrap the rank and percentile filters

jakirkham committed:
Expose rank and percentile filters at top

jakirkham committed:
Add tests for the rank and percentile filters too

jakirkham commented:
Would be curious to know if you have any thoughts on this, @mrocklin.

mrocklin commented:
Is there anything in particular that you would like me to comment on?  Fro a cursory glance it looks like you were able to compose the scipy functions with `da.map_overlap` decently happily.

jakirkham commented:
Anything that comes to mind really.

I tried to approach this in a similar way as was done with `dask.array.fft`. Potential concerns could be the degree of magic and whether it is too much. Another concern could be whether the test coverage is really good enough. Are there ways it could be consolidated? Also I found an oddity between some of the SciPy functions in terms of how they handled the `origin` argument. Do you have any thoughts on that point?

mrocklin commented:
Honestly I've been sufficiently busy that I haven't given this problem enough time to have developed  informed opinions about it.  

We might consider reaching out to some other people to get their thoughts.  @jni comes to mind as potentially interested in this sort of thing.

jakirkham commented on dask_ndfilters/order.py:
```diff
@@ -0,0 +1,224 @@
+    # Workaround for a bug in Dask with 0 depth.
+    #
+    # ref: https://github.com/dask/dask/issues/2258
+    #
```
xref: https://github.com/dask/dask/issues/2258

jakirkham replied:
Should be fixed by PR ( https://github.com/dask/dask/pull/2259 ), which has already been merged.

jakirkham committed:
Improve enforcement of real values for the origin

Make sure any single number is converted into an array. This include
types like complex that are definitely not supported. Once we have an
array, enforce that the type is definitely real. By this we should
capture a few other strange cases like a complex origin array or
similar. To ensure we are raising this error correctly, also include
some tests with unacceptable types for the origin.

jakirkham committed:
Improve enforcement of dimension and type of size

Make sure any single number for size is converted into an appropriate
tuple. Also convert any value for size into a NumPy array. Ensure the
size NumPy array has proper dimensionality and type before finding its
footprint. Also add some tests to ensure these problems are raised to
the user.

jakirkham committed:
Require origin to be of integral type

As only some of the filters handle an origin of floating type by
effectively converting it to an origin of integral type and other
filters simply disallow it, it seems like it would just make more sense
to simply disallow an origin of floating type instead of handling it.
After all this should just be clearer for the end user that they need to
specify a particular index. Allowing a floating type just results in
confusion about how this is handled. If the user has to handle the
floating type themselves, then there is no lack of clarity as to how
this was done. We do this regardless of filter even if the underlying
function allows it. Further we add some tests to verify that floating
type origins are disallowed. Verified that all existing tests of
floating type origin resulted in failures now as well.

jakirkham commented:
I guess if there are no concerns I'll just merge this. That said, this is clearly the very beginning of this library. So there is still plenty of time to discuss changes and improvements to this and later functionality.
